elector/elector.go

```diff
@@ -0,0 +1,289 @@
+package elector
+
+import (
+	"fmt"
+	"path"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/samuel/go-zookeeper/zk"
+)
+
+// IElector is the interface to which the Elector must adhere. Clients may
+// choose to use this, but the Start() method will return a concrete type,
+// keeping in line with 'return concrete types, accept interfaces'.
+type IElector interface {
+	// LeaderIdent returns the current leader of the cluster, or "" if
+	// the current leader is not known.
+	LeaderIdent() string
+
+	// Events returns a channel from which the client should consume events
+	// from the elector.  The channel will be closed after an error event
+	// is sent, as the elector is no longer usable from that point on.
+	Events() <-chan Event
+
+	// Close tidies up any applicable connection details to ZK. Clients
+	// should call then when the elector is no longer needed
+	Close() error
+}
+
+// ensure that Elector adheres to the IElector interface
+var _ IElector = &Elector{}
+
+// Elector handles leadership elections
+type Elector struct {
+	mut         sync.Mutex   //
```

Comment:
i've noticed in other projects that a mutex field appears above all of the mutable state that it's intended to guard. for this struct, applying that same strategy, the only fields below `mut` would be `isLeader` and `leaderIdent` right?

do you think adopting such a strategy would help clarify which variables are guarded by the mutex? in other words:
```
type Elector struct {
  acl
  conn
  events
  ident
  basePath
  closer

  mut // mut guards the following mutable state:
  leaderIdent
  isLeader
}
```

Reply:
i dig it. will add it.